Updates to the Maintainers Guidelines and list of Maintainer Tasks (#80)
hyperledger/tsc: guidelines/MAINTAINERS-guidelines.md

**Emeritus Maintainers**

| Name | GitHub ID | Scope | LFID | Discord ID | Email | Company Affiliation |
|----- | --------- | ----- | ---- | ---------- | ----- | ------------------- |
|      |           |       |      |            |       |                     |

## Maintainer Duties

The `MAINTAINERS.md` file **SHOULD** contain information about the different
maintainer scopes, and for each, the maintainers duties (e.g., maintainers
calls, quarterly reports, code reviews, issue cleansing). See the [Sample
Maintainers](SAMPLE-MAINTAINERS.md) for an example of what to include in this
section for an open source software project. Feel free to evolve that text to
match the needs of the repository and project.

If the repository is for a community specification, or other governance or
documentation purpose, the tasks of the Maintainers (often called "Editors"
in such cases) might be different than for Maintainers of an open source code
project. The Maintainer role is more about approving and merging pull requests
that reflect the agreement of the community, as opposed to code related metrics
(quality, fit for purpose, tests, etc.). In some cases, a GOVERNANCE.md file

> [like this](https://github.com/hyperledger/anoncreds-methods-registry/blob/main/registry/governance.md) might further define the duties of the maintainers.

Reviewer: Please make the link anchor meaningful:
```suggestion
like the [AnonCreds Methods Registry governance.md file](https://github.com/hyperledger/anoncreds-methods-registry/blob/main/registry/governance.md) might further define the duties of the maintainers.
```

nav_order: 2
---
[//]: # (SPDX-License-Identifier: CC-BY-4.0)

## Introduction

> All Hyperledger repositories MUST have a `MAINTAINERS` file (`MAINTAINERS.md` or `MAINTAINERS.rst`) at the top-level directory of the source code. The [SAMPLE-MAINTAINERS.md](SAMPLE-MAINTAINERS.md) can be used as a template by projects creating a new repository.

Reviewer 2: ```suggestion
All Hyperledger repositories **MUST** have a `MAINTAINERS` file (`MAINTAINERS.md` or `MAINTAINERS.rst`) at the top-level directory of the source code. The [SAMPLE-MAINTAINERS.md](SAMPLE-MAINTAINERS.md) can be used as a template by projects creating a new repository.
```

> All Hyperledger repositories **MUST** have a `MAINTAINERS` file (`MAINTAINERS.md` or `MAINTAINERS.rst`) at the top-level directory of the source code. The [SAMPLE-MAINTAINERS.md](SAMPLE-MAINTAINERS.md) can be used as a template by projects creating a new repository.

Reviewer 3: Please make this MAINTAINERS.md _only_.

Reviewer 2: @ryjones : Are you suggesting this as the change?
```suggestion
All Hyperledger repositories **MUST** have a `MAINTAINERS` file (i.e., `MAINTAINERS.md`) at the top-level directory of the source code. The [SAMPLE-MAINTAINERS.md](SAMPLE-MAINTAINERS.md) can be used as a template by projects creating a new repository.
```

Reviewer 3: @tkuhrt yes

Author: Instead, better to just change the generic “MAINTAINERS” references throughout the document to be “MAINTAINERS.md”.  I’ll do that and not have the parentheses in the text here.

  - Each elevated GitHub role needed in a repository requires the Hyperledger Foundation GitHub organization administrators to create and maintain an additional GitHub Team.
  - The Hyperledger Foundation GitHub organization administrators manage a team per Hyperledger Project that includes all contributors to the project. That team is given the "Read" GitHub role in all project repositories. This team need not be documented in the "MAINTAINERS" file.
- Maintainers **MAY** define Maintainers scope within a repository don't require
  elevated GitHub privileges. For example, a scope might include hosting
  community meetings, or contributing to the Hyperledger Project's Quarterly
  report.

If there is more than the single "Maintainer" scope used in a repository, there **MUST** be a list of the repository specific scopes in the MAINTAINERS file. The list must include the scope name, the definition of the scope, and if applicable, the related GitHub Role and Team for the scope.

| Scope | Definition | GitHub Role | GitHub Team |
| ----- | ---------- | ----------- | ----------- |
|       |            |             |             |

## List of Repository Maintainers

> Lists of active and emeritus maintainers MUST be included in the `MAINTAINERS` file.

Reviewer 2: ```suggestion
Lists of active and emeritus maintainers **MUST** be included in the `MAINTAINERS` file.
```

> Changes to the maintainers lists MUST be made via Pull Requests. Once a new `MAINTAINERS` file is created or a PR changing the maintainer lists within the file is merged, a corresponding update to the affected GitHub Teams within the Hyperledger GitHub organization must be made. This is a manual process and the maintainers must ensure that it occurs.

Reviewer 2: ```suggestion
Changes to the maintainers lists **MUST** be made via Pull Requests. Once a new `MAINTAINERS` file is created or a PR changing the maintainer lists within the file is merged, a corresponding update to the affected GitHub Teams within the Hyperledger GitHub organization must be made. This is a manual process and the maintainers must ensure that it occurs.
```

> It is recommended that the lists be sorted alphabetically and contain at least the maintainers name and GitHub ID. If useful, other columns may be added to the table, such as LFID, Discord ID, Email, and Company Affiliation.

Reviewer 2: This is a change from what was previously required. There were reasons for the other fields that I believe are still relevant. Please see my other comment for details.

| ----- | ---------- | ----------- | ----------- |
|       |            |             |             |

## List of Repository Maintainers

Lists of active and emeritus maintainers MUST be included in the `MAINTAINERS` file.

Changes to the maintainers lists MUST be made via Pull Requests. Once a new `MAINTAINERS` file is created or a PR changing the maintainer lists within the file is merged, a corresponding update to the affected GitHub Teams within the Hyperledger GitHub organization must be made. This is a manual process and the maintainers must ensure that it occurs.

It is recommended that the lists be sorted alphabetically and contain at least the maintainers name and GitHub ID. If useful, other columns may be added to the table, such as LFID, Discord ID, Email, and Company Affiliation.

A "Scope" column **MUST** be included if there is more than one scope defined in the repository.

The following shows a suggested format for the information:

**Active Maintainers**

| Maintainer | GitHub ID | LFID | Chat ID | Email | Company Affiliation | Scope |
| ---------- | --------- | ---- | ------- | ----- | ------------------- | ----- |

> |            |           |      |         |       |                     |       |

Reviewer 2: Why are we removing the other fields? The GitHub ID is not always enough to contact the person in question. You are removing the MUST to have chat ID or email. The company affiliation lets us know if we have diversity of maintainers for a project (i.e., that the project is not dependent on a single company to survive).

Author: My experience with creating the table is that the information is hard to come by when initially populating the file when adding it to an existing repository.  It’s fine when starting from scratch, but painful when all you to start from is the GitHub Team.  I also think that some of the fields (e.g., LFID) are unnecessary.

AFAIK — the requirement to be able to contact the person is satisfied by the GitHub Id, so the extra methods were redundant.

I had never seen the company affiliation field previously (my bad 😄 ) and thought that would be painful to maintain. Is it being actively used for diversity reporting or is that asperational?

I’m open to adding some of the fields back, but recommend that we have a good reason for adding each one.

Reviewer 4: I think the idea is that the maintainers themselves could populate this information, so I'm not sure it would be difficult to accumulate.  Any active maintainer should have no problems adding themselves to the table.

The LFID is useful for keeping track of people across companies and projects.  More importantly, it's useful for the people themselves, so they can show others what they've done even if they switch jobs or projects.

We are also actively using company affiliation to measure diversity--it is one of the most important contributor diversity metrics.  Lots of people ask us about the breakdowns of contributors by company, and it makes it difficult for us to answer this question if people contribute using their non-company email addresses (which is fine--we, and potential contributors, just want a way to measure company diversity).

Reviewer 2: With the re-addition of these columns, we still have an open item about whether an email address or chat ID should be required. Currently line 56 says that it is recommended.
